rls/src/main/java/io/grpc/rls/internal/LinkedHashLruCache.java

```diff
@@ -0,0 +1,385 @@
+    this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    delegate = new LinkedHashMap<K, SizedValue>(
+        // rough estimate or minimum hashmap default
+        Math.max((int) (maxEstimatedSizeBytes / 1000), 16),
+        /* loadFactor= */ 0.75f,
+        /* accessOrder= */ true) {
+      @Override
+      protected boolean removeEldestEntry(Map.Entry<K, SizedValue> eldest) {
+        if (estimatedSizeBytes.get() <= LinkedHashLruCache.this.estimatedMaxSizeBytes) {
+          return false;
+        }
+
+        // first, remove at most 1 expired entry
+        boolean removed = cleanupExpiredEntries(1, timeProvider.currentTimeNanos());
+        // handles size based eviction if necessary no expired entry
+        boolean shouldRemove =
+            !removed && shouldInvalidateEldestEntry(eldest.getKey(), eldest.getValue().value);
+        if (shouldRemove) {
+          // remove entry by us to make sure lruIterator and cache is in sync
+          LinkedHashLruCache.this.invalidate(eldest.getKey(), EvictionType.SIZE);
+        }
+        return false;
+      }
+    };
+    checkNotNull(ses, "ses");
+    checkState(cleaningInterval > 0, "cleaning interval must be positive");
+    checkNotNull(cleaningIntervalUnit, "cleaningIntervalUnit");
+    periodicCleaner = new PeriodicCleaner(ses, cleaningInterval, cleaningIntervalUnit).start();
+  }
+
+  /**
+   * Determines if the eldest entry should be kept or not when the cache size limit is reached. Note
+   * that LruCache is access level and the eldest is determined by access pattern.
+   */
+  @SuppressWarnings("unused")
+  protected boolean shouldInvalidateEldestEntry(K eldestKey, V eldestValue) {
+    return true;
+  }
+
+  /** Determines if the entry is already expired or not. */
+  protected abstract boolean isExpired(K key, V value, long nowNanos);
+
+  /**
+   * Returns estimated size of entry to keep track. If it always returns 1, the max size bytes
+   * behaves like max number of entry (default behavior).
+   */
+  @SuppressWarnings("unused")
+  protected int estimateSizeOf(K key, V value) {
+    return 1;
+  }
+
+  /** Updates size for given key if entry exists. It is useful if the cache value is mutated. */
+  public void updateEntrySize(K key) {
+    SizedValue entry = readInternal(key);
+    if (entry == null) {
+      return;
+    }
+    int prevSize = entry.size;
+    int newSize = estimateSizeOf(key, entry.value);
+    entry.size = newSize;
+    estimatedSizeBytes.addAndGet(newSize - prevSize);
+  }
+
+  @Override
+  @Nullable
+  public final V cache(K key, V value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    SizedValue existing;
+    int size = estimateSizeOf(key, value);
+    synchronized (lock) {
+      estimatedSizeBytes.addAndGet(size);
+      existing = delegate.put(key, new SizedValue(size, value));
+      if (existing != null) {
+        evictionListener.onEviction(key, existing, EvictionType.REPLACED);
```

Comment:
it depends on what user expects (in this case me, because it is internal package private api). EXPIRED and REPLACED for the same key in different order can be critical depends on the client's expectation. In order with delayed can be harmful as well.
I borrowed the behavior from guava cache: https://github.com/google/guava/wiki/CachesExplained#removal-listeners

Reply:
i see the concern, the callback acquiring another lock is violating the contract since it is blocking operation (this is why the javadoc asks non blocking operation). but there is nothing prevents it from doing it. we could have a utility class like guava to offload blocking, but it seems overkill for this helper class (offloading option is mentioned in the javadoc).